File: packages/engine/crisp/waveform.py
"""Build a compact waveform summary for the UI: a few dozen peak buckets over the
original audio, plus a flag per bucket for whether that slice was cut. The desktop
app renders this so the user *sees* what Crisp removed.

Reuses the analysis WAV the pipeline already extracted (mono 16 kHz PCM), so it
costs no extra decode. Pure stdlib; non-essential, so any failure yields an empty
summary rather than breaking the clean.
"""

import wave
from array import array
from pathlib import Path

# Cap samples scanned per bucket so a long recording stays fast (peaks over a
# strided scan look identical at UI resolution).
_MAX_SCAN_PER_BUCKET = 300


def _peaks_from_samples(samples, buckets):
    """Normalized (0..1) peak amplitude for each of `buckets` equal slices."""
    n = len(samples)
    if n == 0 or buckets <= 0:
        return []
    peaks = []
    for i in range(buckets):
        lo = (i * n) // buckets
        hi = max(lo + 1, ((i + 1) * n) // buckets)
        step = max(1, (hi - lo) // _MAX_SCAN_PER_BUCKET)
        peak = 0
        j = lo
        while j < hi:
            a = samples[j]
            a = -a if a < 0 else a
            if a > peak:
                peak = a
            j += step
        peaks.append(round(peak / 32768.0, 4))
    return peaks


def _removed_flags(buckets, duration, keep_segments):
    """True for each bucket whose center time falls outside every kept segment —
    i.e. the slices Crisp cut out."""
    if buckets <= 0 or duration <= 0:
        return []
    flags = []
    for i in range(buckets):
        t = (i + 0.5) * duration / buckets
        kept = any(s <= t <= e for s, e in keep_segments)
        flags.append(not kept)
    return flags


def waveform_summary(wav_path, duration, keep_segments, buckets=120):
    """Read the analysis WAV and return {"peaks": [...], "removed": [bool, ...]}.
    Returns empties on any problem — the waveform is a nicety, never load-bearing."""
    try:
        with wave.open(str(Path(wav_path)), "rb") as w:
            if w.getsampwidth() != 2:
                return {"peaks": [], "removed": []}
            samples = array("h")
            samples.frombytes(w.readframes(w.getnframes()))
        return {
            "peaks": _peaks_from_samples(samples, buckets),
            "removed": _removed_flags(buckets, duration, keep_segments),
        }
    except Exception:
        # The waveform is a nicety; whatever goes wrong (unreadable WAV, an odd
        # format, even MemoryError on an hours-long file) must never fail the clean.
        return {"peaks": [], "removed": []}

File: packages/CrispKit/Sources/CrispKit/Waveform.swift
import Foundation

/// The compact waveform summary the UI renders, ported from `crisp/waveform.py`:
/// a few dozen peak buckets over the original audio, plus a flag per bucket for
/// whether that slice was cut, so the user *sees* what Crisp removed.
///
/// Python read these from the analysis WAV it had already written to disk. On iOS
/// there is no such file, so this takes the PCM samples directly: the media layer
/// hands over the same 16 kHz mono buffer it decoded for silence detection, and the
/// waveform costs nothing extra.
public enum Waveform {
    /// Cap samples scanned per bucket so a long recording stays fast. Peaks over a
    /// strided scan look identical at UI resolution.
    static let maxScanPerBucket = 300

    public struct Summary: Hashable, Sendable, Codable {
        /// Normalized 0...1 peak amplitude per bucket.
        public var peaks: [Double]
        /// True for a bucket whose centre falls outside every kept segment.
        public var removed: [Bool]

        public init(peaks: [Double] = [], removed: [Bool] = []) {
            self.peaks = peaks
            self.removed = removed
        }

        public static let empty = Summary()
    }

    /// Normalized peak amplitude for each of `buckets` equal slices of `samples`.
    ///
    /// Samples are signed 16-bit, so the divisor is 32768 and the result is rounded to
    /// four places, matching the Python output byte for byte in the NDJSON payload.
    public static func peaks(from samples: [Int16], buckets: Int) -> [Double] {
        let n = samples.count
        guard n > 0, buckets > 0 else { return [] }

        return (0..<buckets).map { index in
            let lo = (index * n) / buckets
            let hi = max(lo + 1, ((index + 1) * n) / buckets)
            let step = max(1, (hi - lo) / maxScanPerBucket)
            var peak = 0
            var j = lo
            while j < hi {
                // Negate rather than use magnitude: Int16.min has no positive
                // counterpart, so abs() would trap on a full-scale negative sample.
                let sample = Int(samples[j])
                let amplitude = sample < 0 ? -sample : sample
                if amplitude > peak { peak = amplitude }
                j += step
            }
            return (Double(peak) / 32768.0 * 10000).rounded() / 10000
        }
    }

    /// True for each bucket whose centre time falls outside every kept segment, i.e.
    /// the slices Crisp cut out.
    public static func removedFlags(buckets: Int, duration: Double,
                                    keep: [TimeSpan]) -> [Bool] {
        guard buckets > 0, duration > 0 else { return [] }
        return (0..<buckets).map { index in
            let t = (Double(index) + 0.5) * duration / Double(buckets)
            return !keep.contains { $0.contains(t) }
        }
    }

    /// Build the full summary. Non-essential by design: a caller that cannot supply
    /// samples passes an empty array and gets `.empty` rather than an error, because a
    /// missing waveform must never fail a clean.
    public static func summary(samples: [Int16], duration: Double,
                               keep: [TimeSpan], buckets: Int = 120) -> Summary {
        Summary(peaks: peaks(from: samples, buckets: buckets),
                removed: removedFlags(buckets: buckets, duration: duration, keep: keep))
    }
}
